Migrate as much as possible to pip and requirements.txt

JamieMagee opened the pull request:
I've moved all the dependencies from lib/ to requirements.txt. This means it's much easier to keep dependencies up to date, and for developers to work on headphones.

I also added a release script to .travis.yml so that when you create a tag in git, travis will download all the dependencies and package everything into a zip file. From the point of view of a user nothing will change.

rembo10 commented:
Is there any way to pull in specific versions of packages? Also - we modified the musicbrainz package a bit to work with custom mirrors.... It's on my fork of the musicbrainz-ngs repo (and probably needs to be updated.....). Those are the only two problems I forsee with this

JamieMagee commented:
Yeah, in the requirements file you can list the version as well as the dependency name like
`SomePackage==1.2.3` and then to update a package all you need to do is change the version number in `requirements.txt`

I can point the musicbrainz dependency to your github repo and pip will install it, instead of the upstream.

JamieMagee commented:
I've rebased my branch on top of your current develop branch, added versions to requirements.txt, and added your fork of musicbrainzngz over the official one.

rembo10 commented:
If people are running the git version though, they'll have to manually run 'pip install -r requirements.txt' though, correct?

satreix commented:
Yes. If they want to do it the _clean_ way they will have to create a virtualenv and pull you dependencides: 

```
virtualenv -p python2 venv
source venv/bin/activate
pip install -r requirements.txt
```

That is the way to go and it adds 2 nice things:

- no requirements code in headphones repo
- possibility to run several version of dependencies

Is everything ok to merge ?

rembo10 commented:
I'll tell you the only reason why I'm hesitant. The codebase is kind of a
mess. Eventually it needs to be rewritten - even if it's just for python 3.
I've been working on a node version that's a lot cleaner (and uses npm to
install all the external libs). It just seems like kind of a major overhaul
to do at the moment.... I could be wrong though
On Dec 5, 2015 2:56 AM, "Steve Barrau" <notifications@github.com> wrote:

> Yes. If they want to do it the *clean* way they will have to create a
> virtualenv and pull you dependencides:
>
> virtualenv -p python2 venv
> source venv/bin/activate
> pip install -r requirements.txt
>
> That is the way to go and it adds 2 nice things:
>
>    - no requirements code in headphones repo
>    - possibility to run several version of dependencies
>
> Is everything ok to merge ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-162172445>.
>

satreix commented:
> I'll tell you the only reason why I'm hesitant. The codebase is kind of a
> mess. Eventually it needs to be rewritten - even if it's just for python 3.

We are here to help (and I would love a modern python3 version of headphones).

> I've been working on a node version that's a lot cleaner 

How is changing language fixes any code issue ?

> It just seems like kind of a major overhaul to do at the moment....

This pull request does all the work already. Are all the tests passing ?

JamieMagee commented:
I actually had a look into converting headphones to python3 as well. Most of the dependencies support python3, but the main blocker is beets, which is still python2 only.

I agree with @satreix that changing to node, just for the sake of it, won't solve any problems. I'm happy to devote some time to helping headphones going forward.

rembo10 commented:
I was looking into switching because then it can be just Javascript on the
front end and backend. And express makes it pretty easy to write apis. And
dependency management with npm is good. And it's async by nature.

Beets was also the problem with this approach, we'd probably just need to
call it from the command line for now (which were already doing with ffmpeg
and stuff)

The current codebase is just so messy and badly written and repetitive that
I think a rewrite is in order.

Thoughts?
On Dec 7, 2015 2:35 AM, "Jamie Magee" <notifications@github.com> wrote:

> I actually had a look into converting headphones to python3 as well. Most
> of the dependencies support python3, but the main blocker is beets, which
> is still python2 only.
>
> I agree with @satreix <https://github.com/satreix> that changing to node,
> just for the sake of it, won't solve any problems. I'm happy to devote some
> time to helping headphones going forward.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-162477991>.
>

satreix commented:
**Migrating dependencies to pip is free** and will remove a lot of unnecessary code from this repository. I am not to much of a fan of writing the thing in javascript, I'de much rather help on the python version that we have.

I would advocate:

- going _requirements.txt_ all the way first;
- then clean the python code;
- then migrate it to py3.

rembo10 commented:
Why are you opposed to javascript out of curiosity?

I'm not 100% opposed to the idea of using pip, just not sure if it makes
sense to add another layer of complexity to the installation procedure at
this point (although with a new version it would make more sense to keep
the libs out of it).

My idea was to strip the server side down to the bare bones. So like,
rather than keeping a local database that syncs with musicbrainz, just pull
all that info when we need it client side. So we can cut down on a lot of
the cruft, and it would speed everything up by just saving essential stuff
to the db.
On Dec 7, 2015 2:43 PM, "Steve Barrau" <notifications@github.com> wrote:

> *Migrating dependencies to pip is free* and will remove a lot of
> unnecessary code from this repository. I am not to much of a fan of writing
> the thing in javascript, I'de much rather help on the python version that
> we have.
>
> I would advocate:
>
>    - going *requirements.txt* all the way first;
>    - then clean the python code;
>    - then migrate it to py3.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-162692660>.
>

satreix commented:
> Why are you opposed to javascript out of curiosity?

I really do not like js, cant read its mess and shurrely do not like to go anywhere near it. Python feels the other way around.

> not sure if it makes sense to add another layer of complexity

How is pip a layer of complexity? It cleans up your code, and ties your package with the whole python package infrastructure.

> rather than keeping a local database that syncs with musicbrainz, just pull
> all that info when we need it client side. So we can cut down on a lot of the cruft

Server side only right? Just moving that code client side will not make it go away. Just execute for each clients instead of once, on the server. Not bashing, just trying to understand :smile:

JamieMagee commented:
I think it would be just as easy to tidy up what is currently in headphones, rather than start from scratch and take about 6 months to get feature parity with what there is now. 
> I'm not 100% opposed to the idea of using pip, just not sure if it makes sense to add another layer of complexity to the installation procedure at this point

Using `pip` and `requirements.txt` will be exactly the same as using `npm` and `package.json` on node.

I reckon if you:
* Implemented continuous integration (like Travis-CI)
* Transferred this repo to an organisation (Someone is already trying to do thisat https://github.com/headphones)
* Created a website (like https://sonarr.tv/)

You could get more users and developers.

satreix commented:
So, can that PR be merged ? And check whether Travis is happy with it ?

rembo10 commented:
Yeah I'll merge this in and set up travis. Sorry for the delay

rembo10 commented:
Ok, so I've set up travis-ci and the build (develop branch) is passing. So I think I'll need to update the readme to include the new installation instructions, and include the travis-ci status badge...... so is [this](https://github.com/rembo10/headphones/pull/2438#issuecomment-162172445) the best way to install?

satreix commented:
This is definitly the best way to develop on the project.

For installation I guess it should work as well. You'll end up having a few _venv_ but at least you can set your own depencies and versions for cheap.

satreix commented:
@JamieMagee you rock! :+1: :sparkles:
Thanks @rembo10 for the merge.

rembo10 commented:
I'll try to get everything into master tomorrow. Just been pretty busy with
the holidays
On Jan 6, 2016 12:44 PM, "Steve Barrau" <notifications@github.com> wrote:

> @JamieMagee <https://github.com/JamieMagee> you rock! [image: :+1:] [image:
> :sparkles:]
> Thanks @rembo10 <https://github.com/rembo10> for the merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-169315863>.
>

JamieMagee commented:
@satreix Any time :)
@rembo10 You need to create a GitHub OAuth token, encrypt it, and add it to .travis.yml to enable automatic releases from Travis CI when you make a git tag. See [this page](https://docs.travis-ci.com/user/deployment/releases) for info.

rembo10 commented:
Oh ok cool, thanks for the heads up. I'll have to update the Readme with
the new installation instructions as well
On Jan 6, 2016 1:28 PM, "Jamie Magee" <notifications@github.com> wrote:

> @satreix <https://github.com/satreix> Any time :)
> @rembo10 <https://github.com/rembo10> You need to create a GitHub OAuth
> token, encrypt it, and add it to .travis.yml to enable automatic releases
> from Travis CI when you make a git tag. See this page
> <https://docs.travis-ci.com/user/deployment/releases> for info.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-169322711>.
>

satreix commented:
@rembo10 while you are at it you could enable https://requires.io/ for the repository (and add the corresponding badge to the `README.md`) to lookup required python package update if you are ok with it. I find it pretty useful.

rembo10 commented:
If we're keeping the libs out of the repository, does that mean that the user will have to manage the dependencies? Like everytime they update to the latest version, they'll have to update the libs (if we've updated requirements.txt)? I just liked the idea of everything being self-contained and ready to go, so as soon as it was downloaded they can run Headphones.py and it starts up......

JamieMagee commented:
End users shouldn't be running the development version, that's why they're having the issues.

If you create a tag in git on the master branch, Travis will automatically bundle all the libs, create a zip file with everything needed to run headphones and upload it to GitHub as a release.

rembo10 commented:
Yeah, I guess that's a fair point
On Jan 9, 2016 6:03 PM, "Jamie Magee" <notifications@github.com> wrote:

> End users shouldn't be running the development version, that's why they're
> having the issues.
>
> If you create a tag in git on the master branch, Travis will automatically
> bundle all the libs, create a zip file with everything needed to run
> headphones and upload it to GitHub as a release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-170265808>.
>

rembo10 commented:
Hey, sorry for dragging my feet on this. I just want to be careful that it
doesn't break everyone's installs.

Would it be a decent suggestion to install the dependencies into the lib
folder, since we're importing everything from there anyway?

Also, what about the built-in updater? Should we switch to fetching the
Travis builds?
On Jan 9, 2016 6:04 PM, "Remy Varma" <remyvarma@gmail.com> wrote:

> Yeah, I guess that's a fair point
> On Jan 9, 2016 6:03 PM, "Jamie Magee" <notifications@github.com> wrote:
>
>> End users shouldn't be running the development version, that's why
>> they're having the issues.
>>
>> If you create a tag in git on the master branch, Travis will
>> automatically bundle all the libs, create a zip file with everything needed
>> to run headphones and upload it to GitHub as a release.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/rembo10/headphones/pull/2438#issuecomment-170265808>.
>>
>

rembo10 commented:
But even if the user is running the master branch, won't they also have to deal with managing dependencies??

JamieMagee commented:
Yeah, I forgot that the built in updater just pulls from the master branch. The zip files generated by travis will come complete with the libs folder, so the updater would have to check for a new zip folder, download and unzip it.

satreix commented:
Is pip abandoned ? :cry:

satreix commented:
@JamieMagee @rembo10 what happened here? Are we dropping the idea? :cry:

JamieMagee commented:
@satreix I haven't had a lot of time recently, and this requires changing the updater to pull from the packaged file generated by Travis CI, instead of directly from Github.

When this PR was originally merged into develop, it broke a lot of things for people using develop - however I argue that if you're not a developer, but you're relying on develop, you're doing something wrong.

satreix commented:
2016-03-21 6:29 GMT-07:00 Jamie Magee <notifications@github.com>:
> this requires changing the updater to pull from the packaged file generated by Travis CI, instead of directly from Github.

Why do we need to pull from a packaged version if you can make one
yourself? I mean if your system does not support `pip` (running on
some shady tv box or something) you could just script your way into
doing what `pip` does: use `curl` and bash to setup the _lib_
directory for your new version. Right?

> When this PR was originally merged into develop, it broke a lot of things for people using develop.

Ok, that make sense to remove it for now then. Can we do anything to
make the transition smoother?

rembo10 commented:
Hey, sorry for the late reply. I guess it's just a matter of providing a
smooth upgrade path......
On Mar 21, 2016 8:04 PM, "Steve Barrau" <notifications@github.com> wrote:

> 2016-03-21 6:29 GMT-07:00 Jamie Magee <notifications@github.com>:
> > this requires changing the updater to pull from the packaged file
> generated by Travis CI, instead of directly from Github.
>
> Why do we need to pull from a packaged version if you can make one
> yourself? I mean if your system does not support `pip` (running on
> some shady tv box or something) you could just script your way into
> doing what `pip` does: use `curl` and bash to setup the _lib_
> directory for your new version. Right?
>
> > When this PR was originally merged into develop, it broke a lot of
> things for people using develop.
>
> Ok, that make sense to remove it for now then. Can we do anything to
> make the transition smoother?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rembo10/headphones/pull/2438#issuecomment-199451793>
>

satreix commented:
## Ideas

- provide a contrib shell script **for systems that do not have `pip`** that:
    - installs pip locally (python-sphinx has this in their Makefile),
    - use it to get all the unmodified requirements in a _lib_ to replicate current behavior.

This will allow us the move away from having the _lib_ directory in this repo.

- we should **move the entire project** to gh/headphones/headphones
    - this allows moving modified dependencies to their own repo (gh/headphones/musicbrainz-ngs)
        - makes it easier to update/pr/work on those on the side

If people want to work on the whole project (with the modified dependencies) the can always `mkdir lib` and `git clone DEPENDENCY` in there.

What do you think about this? :smile:
